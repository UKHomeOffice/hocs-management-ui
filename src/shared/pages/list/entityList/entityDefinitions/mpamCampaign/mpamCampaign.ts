export default {
    entityListName: 'MPAM_CAMPAIGNS',
    entityName: 'campaign',
    entityNamePlural: 'campaigns',
    entityNameCapitalised: 'Campaign',
    entityRoute: '/manage-mpam-campaigns',
    messages: {
        LOAD_ENTITIES_ERROR: 'There was an error retrieving campaigns. Please try refreshing the page.',
        AMEND_ENTITY_ERROR_DESCRIPTION: 'Something went wrong while amending the campaign. Please try again.',
        AMEND_ENTITY_SUCCESS: 'The campaign was amended successfully',
        ADD_ENTITY_SUCCESS: 'The campaign was added successfully',
        DUPLICATE_ENTITY_ERROR_DESCRIPTION: 'A campaign with those details already exists',
        ADD_ENTITY_ERROR_DESCRIPTION: 'Something went wrong while adding the campaign. Please try again.'
    }
} as EntityDefinition;