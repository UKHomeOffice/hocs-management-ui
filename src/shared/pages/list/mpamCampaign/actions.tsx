import EntityListItem from 'shared/models/entityListItem';

export type SetItemDetails = {
    type: 'SetItemDetails';
    payload: EntityListItem;
};

export type SetSimpleName = {
    type: 'SetSimpleName';
    payload: string;
};

export type SetTitle = {
    type: 'SetTitle';
    payload: string;
};

export type PopulateCampaigns = {
    payload: EntityListItem[];
    type: 'PopulateCampaigns';
};

export type Action =
    SetItemDetails |
    PopulateCampaigns |
    SetTitle |
    SetSimpleName;
