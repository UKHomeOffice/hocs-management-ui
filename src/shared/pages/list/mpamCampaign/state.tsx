import EntityListItem from '../../../models/entityListItem';

export interface State {
    campaigns: EntityListItem[];
    campaignsLoaded: boolean;
}
