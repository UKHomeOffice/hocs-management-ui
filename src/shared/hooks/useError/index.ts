import React, { Reducer } from 'react';
import PageError from '../../models/pageError';
import { Action } from './actions';
import { reducer } from './reducer';
import { FormError } from '../../models/formError';
import ErrorMessage from 'shared/models/errorMessage';

const useError = (formErrorDescription?: string, formErrorTitle?: string)
    : [
        PageError,
        (formError: FormError) => void,
        () => void,
        (error: ErrorMessage) => void
    ] => {
    const [pageError, dispatch] = React.useReducer<Reducer<PageError, Action>>(reducer, {});

    const addFormError = (formError: FormError) => {
        (formErrorDescription || formErrorTitle) && setMessage({ title: formErrorTitle || '', description: formErrorDescription || '' });
        dispatch({ type: 'AddFormError', payload: formError });
    };
    const clearErrors = () => dispatch({ type: 'ClearError' });
    const setMessage = (errorMessage: ErrorMessage) => {
        dispatch({ type: 'SetDescription', payload: errorMessage.description });
        dispatch({ type: 'SetTitle', payload: errorMessage.title });
    };
    return [pageError, addFormError, clearErrors, setMessage];
};

export default useError;
