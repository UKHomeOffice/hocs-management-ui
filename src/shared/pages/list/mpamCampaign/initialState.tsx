import { State } from './state';

export const initialState: State = {
    campaigns: [],
    campaignsLoaded: false
};
