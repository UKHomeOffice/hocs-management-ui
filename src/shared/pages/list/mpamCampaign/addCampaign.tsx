import React, { Reducer } from 'react';
import { RouteComponentProps } from 'react-router';
import { Link } from 'react-router-dom';
import { object, string } from 'yup';
import Submit from '../../../common/components/forms/submit';
import Text from '../../../common/components/forms/text';
import { ApplicationConsumer } from '../../../contexts/application';
import { createListItem } from '../../../services/entityListService';
import { reducer } from './addCampaignReducer';
import ErrorSummary from '../../../common/components/errorSummary';
import { GENERAL_ERROR_TITLE, ADD_CAMPAIGN_ERROR_DESCRIPTION, ADD_CAMPAIGN_SUCCESS, VALIDATION_ERROR_TITLE, DUPLICATE_CAMPAIGN_DESCRIPTION } from '../../../models/constants';
import useError from '../../../hooks/useError';
import ErrorMessage from '../../../models/errorMessage';
import EntityListItem from '../../../models/entityListItem';
import { validate } from '../../../validation';
import InputEventData from '../../../models/inputEventData';

interface AddUnitProps extends RouteComponentProps {
    csrfToken?: string;
}

const validationSchema = object({
    title: string()
        .required()
        .label('Campaign name')
        .matches(/^[a-zA-Z0-9_,.!? ()&]*$/),
    simpleName: string()
        .required()
        .label('Campaign code')
        .matches(/^[a-zA-Z0-9_,.!? ()&]*$/)
});

const AddCampaign: React.FC<AddUnitProps> = ({ csrfToken, history }) => {

    const [pageError, addFormError, clearErrors, setErrorMessage] = useError('', VALIDATION_ERROR_TITLE);
    const [campaign, dispatch] = React.useReducer<Reducer<EntityListItem, InputEventData>>(reducer, {
        uuid: '',
        title: '',
        simpleName: ''
    });

    const handleSubmit = (event: React.FormEvent) => {
        event.preventDefault();
        clearErrors();
        if (validate(validationSchema, campaign, addFormError)) {
            createListItem(campaign, 'MPAM_CAMPAIGNS').then(() => {
                history.push('/', { successMessage: ADD_CAMPAIGN_SUCCESS });
            }).catch((error) => {
                if (error && error.response && error.response.status === 409) {
                    setErrorMessage(new ErrorMessage(DUPLICATE_CAMPAIGN_DESCRIPTION, VALIDATION_ERROR_TITLE));
                } else {
                    setErrorMessage(new ErrorMessage(ADD_CAMPAIGN_ERROR_DESCRIPTION, GENERAL_ERROR_TITLE));
                }
            });
        }
    };

    return (
        <>
            <div className="govuk-grid-row">
                <div className="govuk-grid-column-two-thirds-from-desktop">
                    <Link to="/manage-mpam-campaigns" className="govuk-back-link">Back</Link>
                    <ErrorSummary
                        pageError={pageError}
                    />
                    <h1 className="govuk-heading-xl">
                        Add Campaign
                    </h1>
                </div>
            </div>
            <div className="govuk-grid-row">
                <div className="govuk-grid-column-one-half-from-desktop">
                    <form action="/api/entity/list/MPAM_CAMPAIGNS" method="POST" onSubmit={handleSubmit}>
                        <input type="hidden" name="_csrf" value={csrfToken} />
                        <Text
                            label="Campaign name"
                            name="title"
                            type="text"
                            updateState={({ name, value }) => dispatch({ name, value })}
                            value={campaign.title}
                        />
                        <Text
                            label="Campaign code"
                            name="simpleName"
                            type="text"
                            updateState={({ name, value }) => dispatch({ name, value })}
                            value={campaign.simpleName}
                        />
                        <Submit />
                    </form>
                </div>
            </div>
        </>
    );
};

const WrappedAddUnit = ({ history, location, match }: RouteComponentProps) => (
    <ApplicationConsumer>
        {({ csrf }) => (
            <AddCampaign csrfToken={csrf} history={history} location={location} match={match} />
        )}
    </ApplicationConsumer>
);

export default WrappedAddUnit;
