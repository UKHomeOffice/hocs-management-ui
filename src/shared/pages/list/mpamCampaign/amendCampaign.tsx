import React, { Reducer, useEffect } from 'react';
import { RouteComponentProps } from 'react-router';
import { Link } from 'react-router-dom';
import { object, string } from 'yup';
import Submit from '../../../common/components/forms/submit';
import Text from '../../../common/components/forms/text';
import { ApplicationConsumer } from '../../../contexts/application';
import { updateListItem, getItemDetails } from '../../../services/entityListService';
import { reducer } from './amendCampaignReducer';
import ErrorSummary from '../../../common/components/errorSummary';
import { GENERAL_ERROR_TITLE, AMEND_CAMPAIGN_ERROR_DESCRIPTION, AMEND_CAMPAIGN_SUCCESS, VALIDATION_ERROR_TITLE, LOAD_CAMPAIGN_ERROR_DESCRIPTION } from '../../../models/constants';
import useError from '../../../hooks/useError';
import ErrorMessage from '../../../models/errorMessage';
import { validate } from '../../../validation';
import { Action } from './actions';
import { State } from './amendCampaignState';

interface MatchParams {
    itemUUID: string;
}
interface AmendCampaignProps extends RouteComponentProps<MatchParams> {
    csrfToken?: string;
}

const validationSchema = object({
    title: string()
        .required()
        .label('New campaign name')
        .matches(/^[a-zA-Z0-9_,.!? ()&]*$/)
});

const AmendCampaign: React.FC<AmendCampaignProps> = ({ csrfToken, history, match }) => {
    const initialState: State = {
        uuid: '',
        title: '',
        originalTitle: '',
        simpleName: ''
    };

    const [pageError, addFormError, clearErrors, setErrorMessage] = useError('', VALIDATION_ERROR_TITLE);
    const [state, dispatch] = React.useReducer<Reducer<State, Action>>(reducer, initialState);

    const { params: { itemUUID } } = match;

    useEffect(() => {
        getItemDetails(itemUUID)
            .then(item => dispatch({ type: 'SetItemDetails', payload: item }))
            .catch(() => {
                setErrorMessage(new ErrorMessage(LOAD_CAMPAIGN_ERROR_DESCRIPTION, GENERAL_ERROR_TITLE));
            });
    }, []);

    const handleSubmit = (event: React.FormEvent) => {
        event.preventDefault();
        clearErrors();
        if (validate(validationSchema, state, addFormError)) {
            updateListItem(state, 'MPAM_CAMPAIGNS').then(() => {
                history.push('/', { successMessage: AMEND_CAMPAIGN_SUCCESS });
            }).catch((error) => {
                setErrorMessage(new ErrorMessage(AMEND_CAMPAIGN_ERROR_DESCRIPTION, GENERAL_ERROR_TITLE));
            });
        }
    };

    return (
        <>
            <div className="govuk-grid-row">
                <div className="govuk-grid-column-two-thirds-from-desktop">
                    <Link to="/manage-mpam-campaigns" className="govuk-back-link">Back</Link>
                    <ErrorSummary
                        pageError={pageError}
                    />
                    <h1 className="govuk-heading-xl">
                        Amend Campaign
                    </h1>
                </div>
            </div>
            <div className="govuk-grid-row">
                <div className="govuk-grid-column-one-half-from-desktop">
                    <h3 className="govuk-heading-l">
                        {`Campaign name: ${state.originalTitle}`}
                    </h3>
                    <form action="/api/entity/list/update/MPAM_CAMPAIGNS" method="POST" onSubmit={handleSubmit}>
                        <input type="hidden" name="_csrf" value={csrfToken} />
                        <Text
                            label="New campaign name"
                            name="title"
                            type="text"
                            updateState={({ value }) => dispatch({ type: 'SetTitle', payload: value as string })}
                            value={state.title}
                        />
                        <Text
                            label="Campaign code"
                            name="simpleName"
                            type="text"
                            disabled={true}
                            updateState={({ value }) => dispatch({ type: 'SetSimpleName', payload: value as string })}
                            value={state.simpleName}
                        />
                        <Submit />
                    </form>
                </div>
            </div>
        </>
    );
};

const WrappedAddUnit = ({ history, location, match }: RouteComponentProps<MatchParams>) => (
    <ApplicationConsumer>
        {({ csrf }) => (
            <AmendCampaign csrfToken={csrf} history={history} location={location} match={match} />
        )}
    </ApplicationConsumer>
);

export default WrappedAddUnit;
