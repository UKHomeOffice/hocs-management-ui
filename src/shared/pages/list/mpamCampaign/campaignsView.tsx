import React, { Reducer, useEffect, Fragment } from 'react';
import { RouteComponentProps } from 'react-router';
import { History } from 'history';
import { getListItems } from '../../../services/entityListService';
import { State } from './state';
import { Action } from './actions';
import { reducer } from './reducer';
import { initialState } from './initialState';
import { GENERAL_ERROR_TITLE, LOAD_CAMPAIGNS_ERROR_DESCRIPTION } from '../../../models/constants';
import ErrorSummary from '../../../common/components/errorSummary';
import ErrorMessage from '../../../models/errorMessage';
import useError from '../../../hooks/useError';
import { Link } from 'react-router-dom';
import EntityListItem from 'shared/models/entityListItem';

interface MatchParams {
    teamId: string;
}

interface TeamMembersProps extends RouteComponentProps<MatchParams> { }

const onAddClick = (history: History) => {
    history.push('/manage-mpam-campaigns/add');
};

const CampaignsView: React.FC<TeamMembersProps> = ({ history, match }) => {

    const [pageError, , clearErrors, setErrorMessage] = useError();
    const [state, dispatch] = React.useReducer<Reducer<State, Action>>(reducer, initialState);

    useEffect(() => {
        getListItems('MPAM_CAMPAIGNS')
            .then((campaigns: EntityListItem[]) => dispatch({ type: 'PopulateCampaigns', payload: campaigns }))
            .catch(() => {
                setErrorMessage(new ErrorMessage(LOAD_CAMPAIGNS_ERROR_DESCRIPTION, GENERAL_ERROR_TITLE));
            });
    }, []);

    const amendCampaign = (campaignUUID: string, event: React.FormEvent) => {
        event.preventDefault();
        clearErrors();
        history.push(`/manage-mpam-campaigns/${campaignUUID}/amend`);

    };

    const DisplayCampaignTable = () => (
        <Fragment>
            {state.campaignsLoaded && (
                <table className="govuk-table">
                    <thead className="govuk-table__head">
                        <tr className="govuk-table__row">
                            <th className="govuk-table__header" scope="col">Campaign name</th>
                            <th className="govuk-table__header" scope="col">Campaign code</th>
                            <th className="govuk-table__header" scope="col">Action</th>
                        </tr>
                    </thead>
                    <tbody className="govuk-table__body">
                        {
                            state.campaigns.map((campaign) => {
                                return (
                                    <tr className="govuk-table__row" key={campaign.uuid}>
                                        <td className="govuk-table__cell">{campaign.title}</td>
                                        <td className="govuk-table__cell">{campaign.simpleName}</td>
                                        <td className="govuk-table__cell">
                                            <a href="#" onClick={event => amendCampaign(campaign.uuid, event)}>Amend</a>
                                        </td>
                                    </tr>
                                );
                            })
                        }
                    </tbody>
                </table>
            )}
        </Fragment>
    );

    return (
        <div className="govuk-form-group">
            <Link to="/" className="govuk-back-link">Back</Link>
            <ErrorSummary
                pageError={pageError}
            />
            <div>
                <h1 className="govuk-heading-xl">View and edit campaigns</h1>
                {
                    state.campaignsLoaded ?
                        <div>
                            <DisplayCampaignTable />
                        </div> :
                        <div>
                            <p className="govuk-body">Loading...</p>
                        </div>
                }
                <button type="submit" className="govuk-button govuk-!-margin-right-1 add-team-members-button" data-module="govuk-button" onClick={() => onAddClick(history)}>Add new campaign</button>
            </div>
        </div>
    );
};

export default CampaignsView;
