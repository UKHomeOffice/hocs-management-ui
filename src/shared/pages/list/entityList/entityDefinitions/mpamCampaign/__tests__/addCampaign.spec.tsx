import React from 'react';
import { match, MemoryRouter } from 'react-router-dom';
import { createBrowserHistory, History, Location } from 'history';
import { act, render, RenderResult, wait, fireEvent, waitForElement } from '@testing-library/react';
import AddEntity from '../../../addEntity';
import * as EntityListService from '../../../../../../services/entityListService';
import { GENERAL_ERROR_TITLE, VALIDATION_ERROR_TITLE }
    from '../../../../../../models/constants';
import EntityListItem from '../../../../../../models/entityListItem';
import * as useError from '../../../../../../hooks/useError';
import mpamCampaign from '../mpamCampaign';

let match: match<any>;
let history: History<any>;
let location: Location;
let mockEntityListItem: EntityListItem;
let wrapper: RenderResult;

const useReducerSpy = jest.spyOn(React, 'useReducer');
const reducerDispatch = jest.fn();
const useErrorSpy = jest.spyOn(useError, 'default');
const addFormErrorSpy = jest.fn();
const clearErrorsSpy = jest.fn();
const setMessageSpy = jest.fn();

const AddCampaign = AddEntity(mpamCampaign);

const renderComponent = () => render(
    <MemoryRouter>
        <AddCampaign history={history} location={location} match={match}></AddCampaign>
    </MemoryRouter>
);

jest.spyOn(EntityListService, 'createListItem').mockImplementation(() => Promise.resolve());

beforeEach(() => {
    history = createBrowserHistory();
    match = {
        isExact: true,
        params: {},
        path: '',
        url: ''
    };

    location = {
        hash: '',
        key: '',
        pathname: '',
        search: '',
        state: {}
    };
    mockEntityListItem = {
        title: '',
        simpleName: '',
        uuid: ''
    };
    useReducerSpy.mockImplementation(() => [mockEntityListItem, reducerDispatch]);
    useErrorSpy.mockImplementation(() => [{}, addFormErrorSpy, clearErrorsSpy, setMessageSpy]);
    history.push = jest.fn();
    reducerDispatch.mockReset();
    addFormErrorSpy.mockReset();
    clearErrorsSpy.mockReset();
    setMessageSpy.mockReset();
    act(() => {
        wrapper = renderComponent();
    });
});

describe('when the mpamCampaign addEntity component is mounted', () => {
    it('should render with default props', async () => {
        expect.assertions(1);

        await wait(() => {
            expect(wrapper.container).toMatchSnapshot();
        });
    });
});

describe('when the name is entered', () => {
    it('should be persisted in the page state', async () => {
        expect.assertions(1);

        const nameElement = await waitForElement(async () => {
            return await wrapper.findByLabelText('Campaign name');
        });

        fireEvent.change(nameElement, { target: { name: 'title', value: '__displayTitle__' } });

        await wait(() => {
            expect(reducerDispatch).toHaveBeenCalledWith({ name: 'title', value: '__displayTitle__' });
        });
    });
});

describe('when the code is entered', () => {
    it('should be persisted in the page state', async () => {
        expect.assertions(1);

        const codeElement = await waitForElement(async () => {
            return await wrapper.findByLabelText('Campaign code');
        });

        fireEvent.change(codeElement, { target: { name: 'simpleName', value: '__Code__' } });

        await wait(() => {
            expect(reducerDispatch).toHaveBeenCalledWith({ name: 'simpleName', value: '__Code__' });
        });
    });
});

describe('when the submit button is clicked', () => {
    describe('and the data is filled in', () => {

        beforeEach(async () => {
            mockEntityListItem.title = '__displayName__';
            mockEntityListItem.simpleName = '__shortCode__';
            const submitButton = await waitForElement(async () => {
                return await wrapper.findByText('Submit');
            });

            fireEvent.click(submitButton);
        });

        describe('and the service call is successful', () => {
            it('should redirect to the home page', async () => {
                expect.assertions(1);

                await wait(() => {
                    expect(history.push).toHaveBeenCalledWith('/', { successMessage: 'The campaign was added successfully' });
                });
            });
            it('should call the begin submit action', async () => {
                expect.assertions(1);

                await wait(() => {
                    expect(clearErrorsSpy).toHaveBeenCalled();
                });
            });
        });

        describe('and the service call fails', () => {
            beforeAll(() => {
                jest.spyOn(EntityListService, 'createListItem').mockImplementationOnce(() => Promise.reject({ response: { status: 500 } }));
            });

            it('should set the error state', () => {
                expect(setMessageSpy).toHaveBeenCalledWith({ description: 'Something went wrong while adding the campaign. Please try again.', title: GENERAL_ERROR_TITLE });
            });
            it('should call the begin submit action', () => {
                expect(clearErrorsSpy).toHaveBeenCalled();
            });
        });
        describe('and the service call fails with a 409', () => {
            beforeAll(() => {
                jest.spyOn(EntityListService, 'createListItem').mockImplementationOnce(() => Promise.reject({ response: { status: 409 } }));
            });

            it('should set the error state', () => {
                expect(setMessageSpy).toHaveBeenCalledWith({ description: 'A campaign with those details already exists', title: VALIDATION_ERROR_TITLE });
            });
        });
    });
    describe('and the data is not filled in', () => {
        beforeEach(async () => {
            const submitButton = await waitForElement(async () => {
                return await wrapper.findByText('Submit');
            });

            fireEvent.click(submitButton);
        });

        it('should call the begin submit action', () => {
            expect(clearErrorsSpy).toHaveBeenCalled();
        });

        it('should set the error state', () => {
            expect(addFormErrorSpy).toHaveBeenNthCalledWith(1, { key: 'title', value: 'The Campaign name is required' });
            expect(addFormErrorSpy).toHaveBeenNthCalledWith(2, { key: 'simpleName', value: 'The Campaign code is required' });
        });
    });
});
