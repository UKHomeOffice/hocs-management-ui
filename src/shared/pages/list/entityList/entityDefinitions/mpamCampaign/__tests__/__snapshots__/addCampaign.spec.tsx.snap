// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when the addCampaign component is mounted should render with default props 1`] = `
<div>
  <div
    class="govuk-grid-row"
  >
    <div
      class="govuk-grid-column-two-thirds-from-desktop"
    >
      <a
        class="govuk-back-link"
        href="/manage-mpam-campaigns"
      >
        Back
      </a>
      <h1
        class="govuk-heading-xl"
      >
        Add Campaign
      </h1>
    </div>
  </div>
  <div
    class="govuk-grid-row"
  >
    <div
      class="govuk-grid-column-one-half-from-desktop"
    >
      <form
        action="/api/entity/list/MPAM_CAMPAIGNS"
        method="POST"
      >
        <input
          name="_csrf"
          type="hidden"
          value=""
        />
        <div
          class="govuk-form-group"
        >
          <label
            class="govuk-label govuk-label--s"
            for="title"
            id="title-label"
          >
            Campaign name
          </label>
          <input
            class="govuk-input"
            id="title"
            name="title"
            type="text"
            value=""
          />
        </div>
        <div
          class="govuk-form-group"
        >
          <label
            class="govuk-label govuk-label--s"
            for="simpleName"
            id="simpleName-label"
          >
            Campaign code
          </label>
          <input
            class="govuk-input"
            id="simpleName"
            name="simpleName"
            type="text"
            value=""
          />
        </div>
        <input
          class="govuk-button"
          type="submit"
          value="Submit"
        />
      </form>
    </div>
  </div>
</div>
`;

exports[`when the mpamCampaign addEntity component is mounted should render with default props 1`] = `
<div>
  <div
    class="govuk-grid-row"
  >
    <div
      class="govuk-grid-column-two-thirds-from-desktop"
    >
      <a
        class="govuk-back-link"
        href="/manage-mpam-campaigns"
      >
        Back
      </a>
      <h1
        class="govuk-heading-xl"
      >
        Add Campaign
      </h1>
    </div>
  </div>
  <div
    class="govuk-grid-row"
  >
    <div
      class="govuk-grid-column-one-half-from-desktop"
    >
      <form
        action="/api/entity/list/MPAM_CAMPAIGNS"
        method="POST"
      >
        <input
          name="_csrf"
          type="hidden"
          value=""
        />
        <div
          class="govuk-form-group"
        >
          <label
            class="govuk-label govuk-label--s"
            for="title"
            id="title-label"
          >
            Campaign name
          </label>
          <input
            class="govuk-input"
            id="title"
            name="title"
            type="text"
            value=""
          />
        </div>
        <div
          class="govuk-form-group"
        >
          <label
            class="govuk-label govuk-label--s"
            for="simpleName"
            id="simpleName-label"
          >
            Campaign code
          </label>
          <input
            class="govuk-input"
            id="simpleName"
            name="simpleName"
            type="text"
            value=""
          />
        </div>
        <input
          class="govuk-button"
          type="submit"
          value="Submit"
        />
      </form>
    </div>
  </div>
</div>
`;
